File: Data Structures & Algorithms/two-integer-sum/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @return {number[]}
     */
    twoSum(nums, target) {
        const seen = {}

        for (let i = 0; i < nums.length; i++) {
            const num = nums[i]
            const secondNumIdx = seen[target - num]
            if (secondNumIdx !== undefined) {
                return [i, secondNumIdx]
            }

            seen[num] = i
        }

        return []
    }
}

File: Data Structures & Algorithms/combination-target-sum/submission-0.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        result = []
        addends, currSum = [], 0

        def inner(i):
            nonlocal currSum
            if currSum == target:
                result.append(addends.copy())
                return
            elif currSum > target:
                return

            for j in range(i, len(nums)):
                num = nums[j]
                addends.append(num)
                currSum += num

                inner(j)

                addends.pop()
                currSum -= num

        
        inner(0)
        return result